### ISSUE TITLE
Game version?

### ISSUE THREAD

**Issue by ololoe** [OrfeasZ/Statman#16]
Could you PLEASE show supported versions direct in readme? It doesn't work  with my localised version. As I understand you are reading some values from memory, seems addresses are wrong as it shows random values.

**Comment by ololoe** [comment]
Additionally, how to run it? It prevents game from starting or starting level, it only works if I start a level and then the program.

**Comment by SuiMachine** [comment]
> Additionally, how to run it? It prevents game from starting or starting level, it only works if I start a level and then the program.

Hitman: Blood Money from Steam... and that's pretty much it. OrfeasZ played around with supporting other games, but I don't think he ever finished it.

**Comment by ololoe** [comment]
As I said, I have a localised version of H: BM and it doesn't work. Any way to add support for it?
Under version I meant exe version.

**Comment by OrfeasZ** [comment]
Like @SuiMachine mentioned, Statman only supports the Steam version of HBM. I'm not sure what you mean when you say you have a "localised" version.

**Comment by ololoe** [comment]
Localized means translated. It's from DVD image, not from steam.
Is there a way to determine the exe version?

**Comment by ololoe** [comment]
So how does this tool work? Does it read some addresses in memory?

**Comment by ololoe** [comment]
Could you at least add support for GOG version?

**Comment by SuiMachine** [comment]
> So how does this tool work? Does it read some addresses in memory?

By injecting a DLL into the game, which either reads directly from addresses or creates detours for existing functions (generally the latter) and communicates with main program.

> Could you at least add support for GOG version? 

https://github.com/OrfeasZ/Statman/issues/3
https://github.com/OrfeasZ/Statman/issues/13

> Or where can I download Steam exe?

https://store.steampowered.com/app/6860/Hitman_Blood_Money/
Since even if someone were to send you exe alone, it won't work.

**Comment by ololoe** [comment]
@SuiMachine would you give me 10€ too? Exe may need some dll, but I don't think they will weight much.
How do you know how this program works? Have you worked on it? Where have these addresses got from?

**Comment by SuiMachine** [comment]
> @SuiMachine would you give me 10€ too? Exe may need some dll, but I don't think they will weight much.

No, because there is rarely a value in giving people stuff for free, as my experience with making widescreen fixes thought me.
And no, even if, I were to - you still need Steam running and game registered on your own account.

> How do you know how this program works? Have you worked on it? Where have these addresses got from?

This is GitHub.... there is source available. I was suppose to port Hitman Statistics that nvillemin (Forsaken) wrote. And knowing OrfeasZ he probably got them from reverse engineering parts of code.

> Any idea how to get fast exact +1 witness?

No.

**Comment by ololoe** [comment]
Source of what? I could find addresses inside this tool, I'm asking how they were found.

**Comment by SuiMachine** [comment]
> Source of what? I could find addresses inside this tool, I'm asking how they were found.

Class and cheat address:
https://github.com/OrfeasZ/Statman/blob/master/StatModules/HM3/Src/HM3/HM3Pointers.cpp#L15
https://github.com/OrfeasZ/Statman/blob/master/StatModules/HM3/Src/HM3/HM3Module.cpp#L93

Functions:
https://github.com/OrfeasZ/Statman/blob/master/StatModules/HM3/Src/HM3/HM3Functions.cpp#L15-L19

**Comment by ololoe** [comment]
As I said, I don't need addresses, I'm asking how they were found.

**Comment by ololoe** [comment]
I tried Artmoney. But you need there to sure reproduce events. I found the address of shots fires by 47, for other fields you'll need a sure way to get a cetrain value.

**Comment by SuiMachine** [comment]
So I tried to add some sort of GOG support using Signature scans with wildcards to identify functions that need to be hooked by Statman (since GOG's exe is actually a cracked Securom executable, so the machine code is still encrypted), but despite hooking what I thought was correct functions it still crashes. It doesn't help that if you try to launch the game with debugger, the remaining Securom checks will successfully prevent you from launching the game.

**Comment by aarmono** [comment]
I have a preliminary build that works with GOG build 1.2.27639 [here](https://github.com/aarmono/Statman/releases/tag/v1.3.4). The LoadScene code is almost certainly wrong because that was the only code I had to actually change that didn't involve just changing addresses, but it does work without any apparent adverse effects. Other than that I had to change very little. 

I did do some refactoring of the code so that the HM3 DLL can (in theory) auto-detect the version of game it's injected into and auto-adjust addresses and the like, but not having any other version of the game than the GOG build I can't test that. If I can get some information about the Steam build (see below) I'll add it and see about doing a Pull Request here to get everything in one place.

If someone is able to disassemble the LoadScene function from the Steam build, I ought to be able to check that against the GOG build and find the correct address where the function is located.

Also if someone with the Steam build is able to use Cheat Engine or whatever to find me the address where the version string "blood-build3-20060601-26089" is located I may be able to complete the code to detect the Steam build as well.
